document/field/value.go

```diff
@@ -390,3 +390,57 @@ func (v Values) Clone() Values {
 	}
 	return cloned
 }
+
+// OptionalType is a type "option" which is either "null" or has a valid type.
+// It is similar to a `*ValueType` from a functionality perspective but has
+// lower GC overhead.
+type OptionalType struct {
```

Comment:
Would it be better to just have a `null` `ValueType` instead of this `HasType` `bool`? Also, if we keep the `bool`, should we rename it to `HasValidType`

Reply:
`NullType` is an actual type (i.e., fields with null values) so don't think we should mix this use case with that. Thought about using `UnknownType` but just feels this is more explicitly.

`HasType` feels more succinct so going to keep it unless you feel strongly.